concent_api/core/payments/report.py

```diff
@@ -0,0 +1,759 @@
+        errors,
+        matched_objects,
+        additional_verification_payment,
+    )
+    check_warning_no_subtask_related_with_additional_verification_payment(
+        warnings,
+        additional_verification_payment.subtask_id,
+    )
+    check_warning_payment_value_differs_from_verification_cost_in_settings(
+        warnings,
+        matched_objects,
+        additional_verification_payment,
+    )
+
+
+def check_error_cannot_match_message_with_any_payment(
+    errors: List[Error],
+    matched_objects: List,
+    pending_response: PendingResponse,
+) -> bool:
+    """ Add error if message cannot be matched with any payment."""
+    if len(matched_objects) == 0:
+        errors.append(
+            Error(
+                matching_error=MatchingError.MESSAGE_NOT_MATCHED_WITH_ANY_PAYMENT,
+                message=f'PendingResponse of type {pending_response.response_type} and pk {pending_response.pk} does not have any matching payment.',
+            )
+        )
+        return True
+    return False
+
+
+def check_error_cannot_match_payment_with_any_message(
+    errors: List[Error],
+    matched_objects: List,
+    payment: Any,
+) -> bool:
+    """ Add error if payment cannot be matched with any message."""
+    if len(matched_objects) == 0:
+        errors.append(
+            Error(
+                matching_error=MatchingError.PAYMENT_NOT_MATCHED_WITH_ANY_MESSAGE,
+                message=f'Payment of type {type(payment)} and tx_hash {payment.tx_hash} does not have any matching messages.',
+            )
+        )
+        return True
+    return False
+
+
+def check_error_more_matching_message_for_payment_than_given_amount(
+    errors: List[Error],
+    matched_objects: List,
+    payment: Any,
+    amount: int = 1,
+) -> None:
+    """ Add error if there are more messages matched with a payment than given amount. """
+    if len(matched_objects) > amount:
+        errors.append(
+            Error(
+                matching_error=MatchingError.MORE_THAN_EXPECTED_MATCHING_MESSAGES_FOR_PAYMENT,
+                message=f'Payment of type {type(payment)} and tx_hash {payment.tx_hash} has {len(matched_objects)} matching messages.',
+            )
+        )
+
+
+def check_error_more_matching_payment_for_message_than_given_amount(
+    errors: List[Error],
+    matched_objects: List,
+    pending_response: PendingResponse,
+    amount: int = 1,
+) -> None:
+    """ Add error if there are more payments matched with a message than given amount. """
+    if len(matched_objects) > amount:
+        errors.append(
+            Error(
+                matching_error=MatchingError.MORE_THAN_ONE_MATCHING_PAYMENT_FOR_MESSAGE,
+                message=f'PendingResponse of type {pending_response.response_type} and pk {pending_response.pk} has {len(matched_objects)} matching payments.',
+            )
+        )
+
+
+def check_error_message_value_is_different_than_payment_value(
```

Comment:
This function is typical inner helper for other function in this file. In java it should be 100% private function, how is it python? Shouldn't the name start with `_` to mark this as private, inner helper instead of public api?

Reply:
In python this is just a convention. You can call/import private and protected function however you like. Some tools (I think Pycharm does it) might complain about it but its only that. I think making things private or public makes sense in object oriented approach - which is AFAIK how Java works - so there are actually no functions in Java. And when implementing a class in Python I would follow this convention. But in functional programming I can't see a logical division into public/protected/private functions, as any of those can be reused outside of current module.